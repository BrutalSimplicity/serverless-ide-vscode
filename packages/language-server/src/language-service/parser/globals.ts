import { ASTNode, PropertyASTNode, ObjectASTNode } from './jsonParser';
import { findProperty, getPropertyNodeValue } from './util';
import {
	GlobalsConfig,
	GlobalConfigItem,
	API,
	FUNCTION,
	SIMPLE_TABLE,
	GlobalKeys
} from '../model/globals';

const GLOBALS_KEY: 'Globals' = 'Globals';

export const getDefaultGlobalsConfig = (): GlobalsConfig => ({
	[API]: {
		...GlobalKeys[API],
		properties: []
	},
	[FUNCTION]: {
		...GlobalKeys[FUNCTION],
		properties: []
	},
	[SIMPLE_TABLE]: {
		...GlobalKeys[SIMPLE_TABLE],
		properties: []
	}
});

const collectGlobalPropertiesFromNode = (
	globalsNode: ObjectASTNode,
	globalsConfig: GlobalsConfig
): GlobalsConfig => {
	globalsNode.getChildNodes().forEach(node => {
		if (node.type === 'property') {
			const propertyNode = node as PropertyASTNode;
			const location = propertyNode.key.location;

			if (location in globalsConfig) {
				const configItem: GlobalConfigItem = globalsConfig[location];
				const propertyObjectNode = getPropertyNodeValue(
					propertyNode,
					String(location)
				);

				if (typeof propertyObjectNode !== 'undefined') {
					propertyObjectNode.getChildNodes().forEach(childNode => {
						if (childNode.type === 'property') {
							const propertyChildNode = childNode as PropertyASTNode;

							configItem.properties.push(
								propertyChildNode.key.location
							);
						}
					});
				}
			}
		}
	});

	return globalsConfig;
};

export const collectGlobals = (rootNode: ASTNode): GlobalsConfig => {
	const globalsConfig: GlobalsConfig = getDefaultGlobalsConfig();

	if (rootNode.type === 'object') {
		const globalsNode = getPropertyNodeValue(
			findProperty(
				rootNode as ObjectASTNode,
				node => node.key.location === GLOBALS_KEY
			),
			GLOBALS_KEY
		);

		if (typeof globalsNode !== 'undefined') {
			return collectGlobalPropertiesFromNode(globalsNode, globalsConfig);
		}
	}

	return globalsConfig;
};
