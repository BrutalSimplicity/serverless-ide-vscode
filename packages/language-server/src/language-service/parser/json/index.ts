import { LanguageSettings } from "./../../model/settings"
import { IProblem, ValidationResult } from "./validation-result"
import { ASTNode, ISchemaCollector, IApplicableSchema } from "./ast-node"
import { JSONSchema } from "../../jsonSchema"

export { ASTNode } from "./ast-node"
export { PropertyASTNode } from "./property-ast-node"
export { StringASTNode } from "./string-ast-node"
export { NumberASTNode } from "./number-ast-node"
export { BooleanASTNode } from "./boolean-ast-node"
export { ObjectASTNode } from "./object-ast-node"
export { ArrayASTNode } from "./array-ast-node"
export { NullASTNode } from "./null-ast-node"
export { ErrorCode } from "./validation-result"

// tslint:disable-next-line: max-classes-per-file
class SchemaCollector implements ISchemaCollector {
	schemas: IApplicableSchema[] = []
	private focusOffset: number
	private exclude: ASTNode
	constructor(focusOffset: number = -1, exclude: ASTNode = null) {
		this.focusOffset = focusOffset
		this.exclude = exclude
	}
	add(schema: IApplicableSchema) {
		this.schemas.push(schema)
	}
	merge(other: ISchemaCollector) {
		this.schemas.push(...other.schemas)
	}
	include(node: ASTNode) {
		return (
			(this.focusOffset === -1 || node.contains(this.focusOffset)) &&
			node !== this.exclude
		)
	}
	newSub(): ISchemaCollector {
		return new SchemaCollector(-1, this.exclude)
	}
}

// tslint:disable-next-line: max-classes-per-file
class NoOpSchemaCollector implements ISchemaCollector {
	get schemas() {
		return []
	}
	add() {
		return
	}
	merge() {
		return
	}
	include() {
		return true
	}
	newSub(): ISchemaCollector {
		return this
	}
}

// tslint:disable-next-line: max-classes-per-file

// tslint:disable-next-line: max-classes-per-file
export class JSONDocument {
	readonly root: ASTNode
	readonly syntaxErrors: IProblem[]

	constructor(root: ASTNode, syntaxErrors: IProblem[]) {
		this.root = root
		this.syntaxErrors = syntaxErrors
	}

	getNodeFromOffset(offset: number): ASTNode {
		return this.root && this.root.getNodeFromOffset(offset)
	}

	getNodeFromOffsetEndInclusive(offset: number): ASTNode {
		return this.root && this.root.getNodeFromOffsetEndInclusive(offset)
	}

	visit(visitor: (node: ASTNode) => boolean): void {
		if (this.root) {
			this.root.visit(visitor)
		}
	}

	configureSettings(parserSettings: LanguageSettings) {
		if (this.root) {
			this.root.setParserSettings(parserSettings)
		}
	}

	validate(schema: JSONSchema): IProblem[] {
		if (this.root && schema) {
			const validationResult = new ValidationResult()
			this.root.validate(
				schema,
				validationResult,
				new NoOpSchemaCollector()
			)
			return validationResult.problems
		}
		return null
	}

	getMatchingSchemas(
		schema: JSONSchema,
		focusOffset: number = -1,
		exclude: ASTNode = null
	): IApplicableSchema[] {
		const matchingSchemas = new SchemaCollector(focusOffset, exclude)
		const validationResult = new ValidationResult()
		if (this.root && schema) {
			this.root.validate(schema, validationResult, matchingSchemas)
		}
		return matchingSchemas.schemas
	}

	getValidationProblems(
		schema: JSONSchema,
		focusOffset: number = -1,
		exclude: ASTNode = null
	) {
		const matchingSchemas = new SchemaCollector(focusOffset, exclude)
		const validationResult = new ValidationResult()
		if (this.root && schema) {
			this.root.validate(schema, validationResult, matchingSchemas)
		}
		return validationResult.problems
	}
}
