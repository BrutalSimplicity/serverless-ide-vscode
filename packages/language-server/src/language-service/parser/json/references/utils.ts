import {
	Reference,
	ReferenceType,
	ReferenceTag
} from "../../../model/references"
import { StringASTNode } from "../string-ast-node"

const parameterRegExp = new RegExp("\\${[^}]*}", "g")

export const getSub = (node: StringASTNode): Reference[] => {
	// rest regexp
	parameterRegExp.lastIndex = 0

	let match: RegExpExecArray
	const references: Reference[] = []
	while (
		(match = parameterRegExp.exec(node.value) as RegExpExecArray) != null
	) {
		// Trim the ${} off of the match
		const referencedKey = match[0].substring(2, match[0].length - 1)
		// skip AWS variables references
		if (!referencedKey.startsWith("AWS::")) {
			const reference = {
				key: referencedKey,
				type: ReferenceType.SUB,
				offset: node.start + ReferenceTag.SUB.length + match.index,
				node
			}
			references.push(reference)
		}
	}
	return references
}

export const getRef = (node: StringASTNode): Reference[] => {
	const referencedKey = node.getValue()

	if (!referencedKey.startsWith("AWS::")) {
		const reference = {
			key: referencedKey,
			type: ReferenceType.REF,
			offset: node.start + ReferenceTag.REF.length,
			node
		}
		return [reference]
	}
	return []
}

export const getGetAtt = (node: StringASTNode): Reference[] => {
	return [
		{
			key: node.value,
			type: ReferenceType.GET_ATT,
			offset: node.start + ReferenceTag.GET_ATT.length,
			node
		}
	]
}

export const getIfFindInMapDependsOn = (
	node: StringASTNode,
	referenceType: ReferenceType
): Reference[] => {
	const referencedKey = node.value
	if (!referencedKey.startsWith("AWS::")) {
		return [
			{
				key: referencedKey,
				type: referenceType,
				offset: node.start,
				node
			}
		]
	}
	return []
}
