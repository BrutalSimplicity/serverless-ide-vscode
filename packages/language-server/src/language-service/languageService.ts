import { IConnection } from "vscode-languageserver"
import {
	CompletionList,
	Diagnostic,
	Position,
	TextDocument
} from "vscode-languageserver-types"
import { LanguageSettings } from "./model/settings"
import { YAMLDocument } from "./parser"
import { YAMLCompletion } from "./services/completion"
import { findDocumentSymbols } from "./services/documentSymbols"
import { YAMLHover } from "./services/hover"
import { JSONSchemaService } from "./services/jsonSchema"
import { YAMLValidation } from "./services/validation"

export interface WorkspaceContextService {
	resolveRelativePath(relativePath: string, resource: string): string
}
/**
 * The schema request service is used to fetch schemas. The result should the schema file comment, or,
 * in case of an error, a displayable error string
 */
export type SchemaRequestService = (uri: string) => Promise<string>

export interface CustomFormatterOptions {
	singleQuote?: boolean
	bracketSpacing?: boolean
	proseWrap?: string
}

export interface LanguageService {
	configure(settings: LanguageSettings): void
	doComplete(
		document: TextDocument,
		position: Position,
		doc: YAMLDocument
	): Promise<CompletionList>
	doValidation(document: TextDocument, yamlDocument): Promise<Diagnostic[]>
	doHover(document: TextDocument, position: Position, doc: YAMLDocument)
	findDocumentSymbols(document: TextDocument, doc: YAMLDocument)
	doResolve(completionItem)
}

export function getLanguageService(
	settings: LanguageSettings,
	connection: IConnection
): LanguageService {
	const schemaService = new JSONSchemaService()

	const completer = new YAMLCompletion(schemaService)
	const hover = new YAMLHover(schemaService)
	const yamlValidation = new YAMLValidation(
		schemaService,
		settings.workspaceRoot,
		connection
	)

	const configure = (newSettings: LanguageSettings) => {
		yamlValidation.configure(newSettings)
		hover.configure(newSettings)
		completer.configure(newSettings)
	}

	return {
		configure,
		doComplete: completer.doComplete.bind(completer),
		doResolve: completer.doResolve.bind(completer),
		doValidation: yamlValidation.doValidation.bind(yamlValidation),
		doHover: hover.doHover.bind(hover),
		findDocumentSymbols
	}
}
