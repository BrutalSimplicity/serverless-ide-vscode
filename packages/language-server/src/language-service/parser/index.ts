"use strict"

import {
	ArrayASTNode,
	ASTNode,
	BooleanASTNode,
	ErrorCode,
	JSONDocument,
	NullASTNode,
	NumberASTNode,
	ObjectASTNode,
	PropertyASTNode,
	StringASTNode
} from "./jsonParser"

import * as nls from "vscode-nls"
const localize = nls.loadMessageBundle()

import { Schema, Type } from "js-yaml"
import * as Yaml from "yaml-ast-parser"

import { GlobalsConfig } from "../model/globals"
import {
	getLineStartPositions,
	getPosition
} from "../utils/documentPositionCalculator"
import { collectGlobals, getDefaultGlobalsConfig } from "./globals"
import { parseYamlBoolean } from "./scalar-type"

export interface Problem {
	message: string
	location: {
		start: number
		end: number
	}
	code: ErrorCode
}

export class SingleYAMLDocument extends JSONDocument {
	root: ASTNode
	errors: Problem[]
	warnings: Problem[]
	globalsConfig: GlobalsConfig
	private lines: number[]

	constructor(lines: number[]) {
		super(null, [])
		this.globalsConfig = getDefaultGlobalsConfig()
		this.lines = lines
		this.root = null
		this.errors = []
		this.warnings = []
	}

	getSchemas(schema, doc, node: ASTNode) {
		const matchingSchemas = []
		doc.validate(schema, matchingSchemas, node.start)
		return matchingSchemas
	}

	getNodeFromOffset(offset: number): ASTNode {
		return this.getNodeFromOffsetEndInclusive(offset)
	}

	private getNodeByIndent = (
		lines: number[],
		offset: number,
		node: ASTNode
	) => {
		const { line, column: indent } = getPosition(offset, this.lines)

		const children = node.getChildNodes()

		// tslint:disable-next-line: no-shadowed-variable
		function findNode(children) {
			for (let idx = 0; idx < children.length; idx++) {
				const child = children[idx]

				const { line: childLine, column: childCol } = getPosition(
					child.start,
					lines
				)

				if (childCol > indent) {
					return null
				}

				const newChildren = child.getChildNodes()
				const foundNode = findNode(newChildren)

				if (foundNode) {
					return foundNode
				}

				// We have the right indentation, need to return based on line
				if (childLine === line) {
					return child
				}
				if (childLine > line) {
					// Get previous
					return idx - 1 >= 0 ? children[idx - 1] : child
				}
				// Else continue loop to try next element
			}

			// Special case, we found the correct
			return children[children.length - 1]
		}

		return findNode(children) || node
	}
}

function recursivelyBuildAst(parent: ASTNode, node: Yaml.YAMLNode): ASTNode {
	if (!node) {
		return
	}

	switch (node.kind) {
		case Yaml.Kind.MAP: {
			const instance = node as Yaml.YamlMap

			const result = new ObjectASTNode(
				parent,
				null,
				node.startPosition,
				node.endPosition
			)

			for (const mapping of instance.mappings) {
				result.addProperty(recursivelyBuildAst(
					result,
					mapping
				) as PropertyASTNode)
			}

			return result
		}
		case Yaml.Kind.MAPPING: {
			const instance = node as Yaml.YAMLMapping
			const key = instance.key

			// Technically, this is an arbitrary node in YAML
			// I doubt we would get a better string representation by parsing it
			const keyNode = new StringASTNode(
				null,
				null,
				true,
				key.startPosition,
				key.endPosition
			)
			keyNode.value = key.value

			const result = new PropertyASTNode(parent, keyNode)
			result.end = instance.endPosition

			const valueNode = instance.value
				? recursivelyBuildAst(result, instance.value)
				: new NullASTNode(
						parent,
						key.value,
						instance.endPosition,
						instance.endPosition
				  )
			valueNode.location = key.value

			result.setValue(valueNode)

			return result
		}
		case Yaml.Kind.SEQ: {
			const instance = node as Yaml.YAMLSequence

			const result = new ArrayASTNode(
				parent,
				null,
				instance.startPosition,
				instance.endPosition
			)

			let count = 0
			for (const item of instance.items) {
				if (item === null && count === instance.items.length - 1) {
					break
				}

				// Be aware of https://github.com/nodeca/js-yaml/issues/321
				// Cannot simply work around it here because we need to know if we are in Flow or Block
				const itemNode =
					item === null
						? new NullASTNode(
								parent,
								null,
								instance.endPosition,
								instance.endPosition
						  )
						: recursivelyBuildAst(result, item)

				itemNode.location = count++
				result.addItem(itemNode)
			}

			return result
		}
		case Yaml.Kind.SCALAR: {
			const instance = node as Yaml.YAMLScalar
			const type = Yaml.determineScalarType(instance)

			// The name is set either by the sequence or the mapping case.
			const name = null
			const value = instance.value

			// This is a patch for redirecting values with these strings to be boolean nodes because its not supported in the parser.
			const possibleBooleanValues = [
				"y",
				"Y",
				"yes",
				"Yes",
				"YES",
				"n",
				"N",
				"no",
				"No",
				"NO",
				"on",
				"On",
				"ON",
				"off",
				"Off",
				"OFF"
			]
			if (
				instance.plainScalar &&
				possibleBooleanValues.indexOf(value.toString()) !== -1
			) {
				return new BooleanASTNode(
					parent,
					name,
					parseYamlBoolean(value),
					node.startPosition,
					node.endPosition
				)
			}

			switch (type) {
				case Yaml.ScalarType.null: {
					return new StringASTNode(
						parent,
						name,
						false,
						instance.startPosition,
						instance.endPosition
					)
				}
				case Yaml.ScalarType.bool: {
					return new BooleanASTNode(
						parent,
						name,
						Yaml.parseYamlBoolean(value),
						node.startPosition,
						node.endPosition
					)
				}
				case Yaml.ScalarType.int: {
					const result = new NumberASTNode(
						parent,
						name,
						node.startPosition,
						node.endPosition
					)
					result.value = Yaml.parseYamlInteger(value)
					result.isInteger = true
					return result
				}
				case Yaml.ScalarType.float: {
					const result = new NumberASTNode(
						parent,
						name,
						node.startPosition,
						node.endPosition
					)
					result.value = Yaml.parseYamlFloat(value)
					result.isInteger = false
					return result
				}
				case Yaml.ScalarType.string: {
					const result = new StringASTNode(
						parent,
						name,
						false,
						node.startPosition,
						node.endPosition
					)
					result.value = node.value
					return result
				}
			}

			break
		}
		case Yaml.Kind.ANCHOR_REF: {
			const instance = (node as Yaml.YAMLAnchorReference).value

			return (
				recursivelyBuildAst(parent, instance) ||
				new NullASTNode(
					parent,
					null,
					node.startPosition,
					node.endPosition
				)
			)
		}
		case Yaml.Kind.INCLUDE_REF: {
			const result = new StringASTNode(
				parent,
				null,
				false,
				node.startPosition,
				node.endPosition
			)
			result.value = node.value
			return result
		}
	}
}

function convertError(e: Yaml.YAMLException): Problem {
	return {
		message: `${e.reason}`,
		location: {
			start: e.mark.position,
			end: e.mark.position + e.mark.column
		},
		code: ErrorCode.Undefined
	}
}

function createJSONDocument(
	yamlDoc: Yaml.YAMLNode,
	startPositions: number[],
	text: string
) {
	const doc = new SingleYAMLDocument(startPositions)
	doc.root = recursivelyBuildAst(null, yamlDoc)
	doc.globalsConfig = collectGlobals(doc.root)

	if (!doc.root) {
		// TODO: When this is true, consider not pushing the other errors.
		doc.errors.push({
			message: localize(
				"Invalid symbol",
				"Expected a YAML object, array or literal"
			),
			code: ErrorCode.Undefined,
			location: { start: yamlDoc.startPosition, end: yamlDoc.endPosition }
		})

		return doc
	}

	const duplicateKeyReason = "duplicate key"

	// Patch ontop of yaml-ast-parser to disable duplicate key message on merge key
	const isDuplicateAndNotMergeKey = (
		error: Yaml.YAMLException,
		yamlText: string
	) => {
		const errorConverted = convertError(error)
		const errorStart = errorConverted.location.start
		const errorEnd = errorConverted.location.end
		if (
			error.reason === duplicateKeyReason &&
			yamlText.substring(errorStart, errorEnd).startsWith("<<")
		) {
			return false
		}
		return true
	}
	doc.errors = yamlDoc.errors
		.filter(e => e.reason !== duplicateKeyReason && !e.isWarning)
		.map(e => convertError(e))
	doc.warnings = yamlDoc.errors
		.filter(
			e =>
				(e.reason === duplicateKeyReason &&
					isDuplicateAndNotMergeKey(e, text)) ||
				e.isWarning
		)
		.map(e => convertError(e))

	return doc
}

// tslint:disable-next-line: max-classes-per-file
export class YAMLDocument {
	documents: SingleYAMLDocument[]

	constructor(documents: SingleYAMLDocument[]) {
		this.documents = documents
	}
}

export const parse = (text: string, customTags = []): YAMLDocument => {
	const startPositions = getLineStartPositions(text)
	// This is documented to return a YAMLNode even though the
	// typing only returns a YAMLDocument
	const yamlDocs = []

	const schemaWithAdditionalTags = Schema.create(
		customTags.map(tag => {
			const typeInfo = tag.split(" ")
			return new Type(typeInfo[0], { kind: typeInfo[1] || "scalar" })
		})
	)

	// We need compiledTypeMap to be available from schemaWithAdditionalTags before we add the new custom properties
	customTags.map(tag => {
		const typeInfo = tag.split(" ")
		schemaWithAdditionalTags.compiledTypeMap[typeInfo[0]] = new Type(
			typeInfo[0],
			{ kind: typeInfo[1] || "scalar" }
		)
	})

	const additionalOptions: Yaml.LoadOptions = {
		schema: schemaWithAdditionalTags
	}

	Yaml.loadAll(text, docItem => yamlDocs.push(docItem), additionalOptions)

	const doc = new YAMLDocument(
		yamlDocs.map(docItem => {
			const jsonDocument = createJSONDocument(
				docItem,
				startPositions,
				text
			)

			return jsonDocument
		})
	)

	return doc
}
