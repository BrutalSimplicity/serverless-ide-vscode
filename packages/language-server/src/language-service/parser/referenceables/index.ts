import { CLOUD_FORMATION, SERVERLESS_FRAMEWORK } from "./../../model/document"
import {
	ASTNode,
	ObjectASTNode,
	PropertyASTNode,
	StringASTNode
} from "./../../parser/json"
import { getNodeItemByKey } from "../../utils/yaml"
import { DocumentType, SAM } from "../../model/document"
import { Referenceables, Referenceable } from "../../model/referenceables"
import { ReferenceEntityType } from "../../model/references"
import { SortedHash } from "../../model/sortedHash"

export const generateEmptyReferenceables = (): Referenceables => ({
	hash: {
		[ReferenceEntityType.PARAMETER]: new SortedHash(),
		[ReferenceEntityType.RESOURCE]: new SortedHash(),
		[ReferenceEntityType.MAPPING]: new SortedHash(),
		[ReferenceEntityType.OUTPUT]: new SortedHash()
	},
	lookup: new WeakMap()
})

const addReferenceable = (
	referenceables: Referenceables,
	entityType: ReferenceEntityType,
	referenceable: Referenceable
) => {
	referenceables.hash[entityType].add(referenceable.id, referenceable)
	referenceables.lookup.set(referenceable.node, referenceable)
}

const getResourceName = (resourceNode: ASTNode): string | void => {
	if (resourceNode) {
		const resourceValue = resourceNode.get(["Type"])

		if (resourceValue instanceof StringASTNode) {
			return resourceValue.value
		}
	}
}

const collectReferenceablesFromNode = (
	entityType: ReferenceEntityType,
	node: PropertyASTNode | void,
	referenceables: Referenceables,
	isSam: boolean = false
): Referenceables => {
	if (node) {
		const { value } = node
		if (value instanceof ObjectASTNode) {
			value.properties.forEach(property => {
				const key = property.key.value
				const resourceName =
					entityType === ReferenceEntityType.RESOURCE
						? getResourceName(property.value)
						: undefined

				const referenceable = {
					id: key,
					node: property,
					entityType,
					resourceName
				}

				addReferenceable(referenceables, entityType, referenceable)

				if (isSam && resourceName === "AWS::Serverless::Function") {
					const roleReferenceable = {
						id: `${key}Role`,
						node: property,
						entityType,
						resourceName: "AWS::IAM::Role"
					}

					// we don't need to add the role into lookup hash
					referenceables.hash[entityType].add(
						roleReferenceable.id,
						roleReferenceable
					)
				}
			})
		}
	}

	return referenceables
}

const collectCfnReferenceables = (
	node: ASTNode,
	isSam: boolean = false
): Referenceables => {
	const referenceables = generateEmptyReferenceables()

	collectReferenceablesFromNode(
		ReferenceEntityType.PARAMETER,
		getNodeItemByKey(node, "Parameters"),
		referenceables
	)

	collectReferenceablesFromNode(
		ReferenceEntityType.RESOURCE,
		getNodeItemByKey(node, "Resources"),
		referenceables,
		isSam
	)

	collectReferenceablesFromNode(
		ReferenceEntityType.MAPPING,
		getNodeItemByKey(node, "Mappings"),
		referenceables
	)

	collectReferenceablesFromNode(
		ReferenceEntityType.OUTPUT,
		getNodeItemByKey(node, "Output"),
		referenceables
	)

	return referenceables
}

const collectServerlessReferenceables = (node: ASTNode): Referenceables => {
	const referenceables = generateEmptyReferenceables()
	const resourcesProperty = getNodeItemByKey(node, "resources")

	if (resourcesProperty && resourcesProperty.value instanceof ObjectASTNode) {
		const node = getNodeItemByKey(resourcesProperty.value, "Resources")

		if (node) {
			collectReferenceablesFromNode(
				ReferenceEntityType.RESOURCE,
				node,
				referenceables
			)
		}
	}

	return referenceables
}

export const collectReferenceables = (
	documentType: DocumentType,
	node: ASTNode
): Referenceables => {
	switch (documentType) {
		case SAM:
			return collectCfnReferenceables(node, true)
		case CLOUD_FORMATION:
			return collectCfnReferenceables(node)
		case SERVERLESS_FRAMEWORK:
			return collectServerlessReferenceables(node)
		default:
			return generateEmptyReferenceables()
	}
}
