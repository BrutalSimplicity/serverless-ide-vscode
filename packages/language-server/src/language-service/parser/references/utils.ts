import { Reference, ReferenceType } from "../../model/references"
import { StringASTNode } from "../json/string-ast-node"

const parameterRegExp = new RegExp("\\${[^}]*}", "g")

export const getSub = (node: StringASTNode): Reference[] => {
	// rest regexp
	parameterRegExp.lastIndex = 0

	let match: RegExpExecArray
	const references: Reference[] = []
	while (
		(match = parameterRegExp.exec(node.value) as RegExpExecArray) != null
	) {
		// Trim the ${} off of the match
		const referencedKey = match[0].substring(2, match[0].length - 1)
		// skip AWS variables references
		if (!referencedKey.startsWith("AWS::")) {
			const reference = {
				key: referencedKey,
				type: ReferenceType.SUB,
				offset: node.start + match.index,
				node
			}
			references.push(reference)
		}
	}
	return references
}

export const getRef = (node: StringASTNode): Reference[] => {
	const referencedKey = node.value

	if (!referencedKey.startsWith("AWS::")) {
		const reference = {
			key: referencedKey,
			type: ReferenceType.REF,
			offset: node.start,
			node
		}
		return [reference]
	}
	return []
}

export const getGetAtt = (node: StringASTNode): Reference[] => {
	// TODO: handle reference properties
	const [referencedKey] = node.value.split(".")

	return [
		{
			key: referencedKey,
			type: ReferenceType.GET_ATT,
			offset: node.start,
			node
		}
	]
}

export const getDependsOn = (node: StringASTNode): Reference[] => {
	const referencedKey = node.value
	if (!referencedKey.startsWith("AWS::")) {
		return [
			{
				key: referencedKey,
				type: ReferenceType.DEPENDS_ON,
				offset: node.start,
				node
			}
		]
	}
	return []
}
