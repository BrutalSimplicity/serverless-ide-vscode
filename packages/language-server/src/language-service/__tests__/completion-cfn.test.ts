import { TextDocument } from "vscode-languageserver"
import { getLanguageService } from "../languageService"
import { parse as parseYAML } from "../parser"
import { completionHelper } from "../utils/completion-helper"

const languageService = getLanguageService([])

const languageSettings = {
	completion: true
}
languageService.configure(languageSettings)

const cfnTemplateWithVersion = `
AWSTemplateFormatVersion: 2010-09-09
`

const cfnTemplateWithResources =
	"AWSTemplateFormatVersion: 2010-09-09\nResources:\n  AWS:Dyna"

describe("Auto Completion Tests", () => {
	describe("doComplete", () => {
		const setup = (content: string) => {
			return TextDocument.create(
				"file://~/Desktop/cfn.yaml",
				"yaml",
				0,
				content
			)
		}

		const parseSetup = (content: string, offset: number) => {
			const testTextDocument = setup(content)
			const position = testTextDocument.positionAt(offset)
			const output = completionHelper(testTextDocument, position)

			return languageService.doComplete(
				testTextDocument,
				output.newPosition,
				parseYAML(output.newText)
			)
		}

		test("should autocomplete on the root node", async () => {
			const content = [
				"AWSTemplateFormatVersion: 2010-09-09",
				"Resources:"
			].join("\n")
			const result = await parseSetup(content, content.length - 1)
			expect(result.items).not.toHaveLength(0)
			expect(result).toMatchSnapshot()
		})

		test.skip("should autocomplete resources", async () => {
			const content = [
				"AWSTemplateFormatVersion: 2010-09-09",
				"Resources:",
				"  AWS::Dyna"
			].join("\n")
			const result = await parseSetup(content, content.length - 1)
			expect(result.items).not.toHaveLength(0)
			expect(result).toMatchSnapshot()
		})
	})
})
