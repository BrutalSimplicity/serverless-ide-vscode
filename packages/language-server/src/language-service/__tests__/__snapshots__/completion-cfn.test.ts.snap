// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auto Completion Tests doComplete should autocomplete on the root node 1`] = `
Object {
  "isIncomplete": false,
  "items": Array [
    Object {
      "documentation": "",
      "insertText": "Conditions:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Conditions",
      "textEdit": Object {
        "newText": "Conditions:
	",
        "range": Object {
          "end": Object {
            "character": 9,
            "line": 1,
          },
          "start": Object {
            "character": 0,
            "line": 1,
          },
        },
      },
    },
    Object {
      "documentation": "Template description",
      "insertText": "Description: $1",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Description",
      "textEdit": Object {
        "newText": "Description: $1",
        "range": Object {
          "end": Object {
            "character": 9,
            "line": 1,
          },
          "start": Object {
            "character": 0,
            "line": 1,
          },
        },
      },
    },
    Object {
      "documentation": "",
      "insertText": "Mappings:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Mappings",
      "textEdit": Object {
        "newText": "Mappings:
	",
        "range": Object {
          "end": Object {
            "character": 9,
            "line": 1,
          },
          "start": Object {
            "character": 0,
            "line": 1,
          },
        },
      },
    },
    Object {
      "documentation": "",
      "insertText": "Metadata:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Metadata",
      "textEdit": Object {
        "newText": "Metadata:
	",
        "range": Object {
          "end": Object {
            "character": 9,
            "line": 1,
          },
          "start": Object {
            "character": 0,
            "line": 1,
          },
        },
      },
    },
    Object {
      "documentation": "",
      "insertText": "Outputs:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Outputs",
      "textEdit": Object {
        "newText": "Outputs:
	",
        "range": Object {
          "end": Object {
            "character": 9,
            "line": 1,
          },
          "start": Object {
            "character": 0,
            "line": 1,
          },
        },
      },
    },
    Object {
      "documentation": "",
      "insertText": "Parameters:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Parameters",
      "textEdit": Object {
        "newText": "Parameters:
	",
        "range": Object {
          "end": Object {
            "character": 9,
            "line": 1,
          },
          "start": Object {
            "character": 0,
            "line": 1,
          },
        },
      },
    },
    Object {
      "documentation": "",
      "insertText": "Resources:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Resources",
      "textEdit": Object {
        "newText": "Resources:
	",
        "range": Object {
          "end": Object {
            "character": 9,
            "line": 1,
          },
          "start": Object {
            "character": 0,
            "line": 1,
          },
        },
      },
    },
    Object {
      "documentation": "",
      "insertText": "Transform:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Transform",
      "textEdit": Object {
        "newText": "Transform:
	",
        "range": Object {
          "end": Object {
            "character": 9,
            "line": 1,
          },
          "start": Object {
            "character": 0,
            "line": 1,
          },
        },
      },
    },
  ],
}
`;

exports[`Auto Completion Tests doComplete should autocomplete resources 1`] = `
Object {
  "isIncomplete": false,
  "items": Array [
    Object {
      "documentation": "",
      "insertText": "Conditions:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Conditions",
      "textEdit": Object {
        "newText": "Conditions:
	",
        "range": Object {
          "end": Object {
            "character": 0,
            "line": 4,
          },
          "start": Object {
            "character": 0,
            "line": 4,
          },
        },
      },
    },
    Object {
      "documentation": "Template description",
      "insertText": "Description: $1",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Description",
      "textEdit": Object {
        "newText": "Description: $1",
        "range": Object {
          "end": Object {
            "character": 0,
            "line": 4,
          },
          "start": Object {
            "character": 0,
            "line": 4,
          },
        },
      },
    },
    Object {
      "documentation": "",
      "insertText": "Mappings:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Mappings",
      "textEdit": Object {
        "newText": "Mappings:
	",
        "range": Object {
          "end": Object {
            "character": 0,
            "line": 4,
          },
          "start": Object {
            "character": 0,
            "line": 4,
          },
        },
      },
    },
    Object {
      "documentation": "",
      "insertText": "Metadata:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Metadata",
      "textEdit": Object {
        "newText": "Metadata:
	",
        "range": Object {
          "end": Object {
            "character": 0,
            "line": 4,
          },
          "start": Object {
            "character": 0,
            "line": 4,
          },
        },
      },
    },
    Object {
      "documentation": "",
      "insertText": "Outputs:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Outputs",
      "textEdit": Object {
        "newText": "Outputs:
	",
        "range": Object {
          "end": Object {
            "character": 0,
            "line": 4,
          },
          "start": Object {
            "character": 0,
            "line": 4,
          },
        },
      },
    },
    Object {
      "documentation": "",
      "insertText": "Parameters:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Parameters",
      "textEdit": Object {
        "newText": "Parameters:
	",
        "range": Object {
          "end": Object {
            "character": 0,
            "line": 4,
          },
          "start": Object {
            "character": 0,
            "line": 4,
          },
        },
      },
    },
    Object {
      "documentation": "",
      "insertText": "Transform:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "Transform",
      "textEdit": Object {
        "newText": "Transform:
	",
        "range": Object {
          "end": Object {
            "character": 0,
            "line": 4,
          },
          "start": Object {
            "character": 0,
            "line": 4,
          },
        },
      },
    },
  ],
}
`;
